go-controller/pkg/ovn/base_network_controller_policy.go

```diff
@@ -1660,28 +1741,66 @@ func (oc *DefaultNetworkController) addHairpinAllowACL() error {
 		match = fmt.Sprintf("(%s || %s)", v4Match, v6Match)
 	}
 
-	ingressACLIDs := oc.getNetpolDefaultACLDbIDs(string(knet.PolicyTypeIngress))
-	ingressACL := BuildACL("", types.DefaultAllowPriority, match,
+	ingressACLIDs := bnc.getNetpolDefaultACLDbIDs(string(knet.PolicyTypeIngress))
+	ingressACL := bnc.BuildACL("", types.DefaultAllowPriority, match,
 		nbdb.ACLActionAllowRelated, nil, lportIngress, ingressACLIDs.GetExternalIDs())
 
-	egressACLIDs := oc.getNetpolDefaultACLDbIDs(string(knet.PolicyTypeEgress))
-	egressACL := BuildACL("", types.DefaultAllowPriority, match,
+	egressACLIDs := bnc.getNetpolDefaultACLDbIDs(string(knet.PolicyTypeEgress))
+	egressACL := bnc.BuildACL("", types.DefaultAllowPriority, match,
 		nbdb.ACLActionAllowRelated, nil, lportEgressAfterLB, egressACLIDs.GetExternalIDs())
 
-	ops, err := libovsdbops.CreateOrUpdateACLsOps(oc.nbClient, nil, ingressACL, egressACL)
+	ops, err := libovsdbops.CreateOrUpdateACLsOps(bnc.nbClient, nil, ingressACL, egressACL)
 	if err != nil {
 		return fmt.Errorf("failed to create or update hairpin allow ACL %v", err)
 	}
 
-	ops, err = libovsdbops.AddACLsToPortGroupOps(oc.nbClient, ops, types.ClusterPortGroupName, ingressACL, egressACL)
+	ops, err = libovsdbops.AddACLsToPortGroupOps(bnc.nbClient, ops, types.ClusterPortGroupName, ingressACL, egressACL)
 	if err != nil {
 		return fmt.Errorf("failed to add ACL hairpin allow acl to port group: %v", err)
 	}
 
-	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+	_, err = libovsdbops.TransactAndCheck(bnc.nbClient, ops)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
+
+// WatchNetworkPolicy starts the watching of multinetworkpolicy resource and calls
+// back the appropriate handler logic
+func (bnc *BaseNetworkController) WatchNetworkPolicy() error {
+	if bnc.IsSecondary() && !config.OVNKubernetesFeature.EnableMultiNetworkPolicy {
+		return nil
+	}
+
+	if bnc.policyHandler != nil {
+		return nil
+	}
+	handler, err := bnc.retryNetworkPolicies.WatchResource()
+	if err != nil {
+		bnc.policyHandler = handler
+	}
+	return err
+}
+
+func cleanupPolicyLogicalEntities(nbClient libovsdbclient.Client, ops []ovsdb.Operation, netName string) ([]ovsdb.Operation, error) {
+	var err error
+	portGroupPredicate := func(item *nbdb.PortGroup) bool {
+		return item.ExternalIDs[types.NetworkExternalID] == netName
+	}
+	ops, err = libovsdbops.DeletePortGroupsWithPredicateOps(nbClient, ops, portGroupPredicate)
+	if err != nil {
+		return ops, fmt.Errorf("failed to get ops to delete port group of network %s", netName)
+	}
+
+	controllerName := netName + "-network-controller"
+	asPredicate := func(item *nbdb.AddressSet) bool {
+		return item.ExternalIDs[libovsdbops.OwnerControllerKey.String()] == controllerName
+	}
+	ops, err = libovsdbops.DeleteAddressSetsWithPredicateOps(nbClient, ops, asPredicate)
```

Comment:
all addresssets belongs to this network.

Reply:
but it actually deletes all address sets and port groups (including multicast port group), so i guess the name is not really correct, maybe you can just put the contents of this function inline to the secondary controller cleanup?
Also, can we please move this function to the secondary controller code, because it is not used by the default network controller.

Reply:
This function is shared by all secondary networks, so I'd rather not to embed them in all secondary network controller cleanup code. any suggestion on naming? multicast support is kind of policy (where or not allow multicast), so I personally think that function name is ok.